src/tactic/omega/nat/main.lean

```diff
@@ -0,0 +1,175 @@
+/-
+Copyright (c) 2019 Seul Baek. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Seul Baek
+
+Main procedure for linear natural number arithmetic.
+-/
+
+import tactic.omega.prove_unsats
+import tactic.omega.nat.dnf
+import tactic.omega.nat.neg_elim
+
+open tactic 
+
+namespace omega
+namespace nat
+
+local notation `&` k    := preterm.cst k
+local infix `**`  : 300 := preterm.var 
+local notation t `+*` s := preterm.add t s
+local notation t `-*` s := preterm.sub t s
+
+local notation x `=*` y := form.eq x y
+local notation x `≤*` y := form.le x y
+local notation `¬*` p   := form.not p
+local notation p `∨*` q := form.or p q
+local notation p `∧*` q := form.and p q
+
+   
+run_cmd mk_simp_attr `sugar_nat 
+attribute [sugar_nat] 
+  not_le not_lt
+  nat.lt_iff_add_one_le
+  nat.succ_eq_add_one
+  or_false false_or
+  and_true true_and
+  ge gt mul_add add_mul mul_comm
+  classical.imp_iff_not_or
+  classical.iff_iff_not_or_and_or_not
+
+meta def desugar := `[try {simp only with sugar_nat}]
+
+lemma univ_close_of_unsat_neg_elim_not (m) (p : form) :
+  (neg_elim (¬* p)).unsat → univ_close p (λ _, 0) m :=  
+begin
+  intro h1, apply univ_close_of_valid,
+  apply valid_of_unsat_not, intro h2, apply h1,
+  apply form.sat_of_implies_of_sat implies_neg_elim h2,
+end
+
+meta def preterm.prove_sub_free : preterm → tactic expr 
```

Comment:
This, and `neg_free`, are straightforward inductions on the terms. I think you can just provide a decidable instance and use `dec_trivial` rather than proof construction here.